hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -0,0 +1,204 @@
+
+    async def delete_object(self, bucket: str, name: str, **kwargs) -> None:
+        await self.delete(f'/b/{bucket}/o/{urllib.parse.quote(name, safe="")}', **kwargs)
+
+    async def _list_objects(self, bucket: str, **kwargs) -> Any:
+        return await self.get(f'/b/{bucket}/o', **kwargs)
+
+
+class GoogleStorageAsyncFS(AsyncFS):
+    def __init__(self, storage_client: Optional[StorageClient] = None):
+        if not storage_client:
+            storage_client = StorageClient()
+        self._storage_client = storage_client
+
+    def schemes(self) -> Set[str]:
+        return {'gs'}
+
+    @staticmethod
+    def _get_bucket_name(url: str) -> Tuple[str, str]:
+        parsed = urllib.parse.urlparse(url)
+        if parsed.scheme != 'gs':
+            raise ValueError(f"invalid scheme, expected gs: {parsed.scheme}")
+
+        name = parsed.path
+        if name:
+            assert name[0] == '/'
+            name = name[1:]
+
+        return (parsed.netloc, name)
+
+    async def open(self, url: str, mode: str = 'r') -> AsyncStream:
+        if not all(c in 'rwxabt+' for c in mode):
+            raise ValueError(f"invalid mode: {mode}")
+        if 't' in mode and 'b' in mode:
+            raise ValueError(f"can't have text and binary mode at once: {mode}")
+        if 'b' not in mode:
+            raise ValueError(f"text mode not supported: {mode}")
+        if 'x' in mode:
+            raise ValueError(f"exclusive creation not supported: {mode}")
+        if 'a' in mode:
+            raise ValueError(f"append mode not supported: {mode}")
+        if '+' in mode:
+            raise ValueError(f"updating not supported: {mode}")
+        if ('r' in mode) + ('w' in mode) != 1:
+            raise ValueError(f"must have exactly one of read/write mode: {mode}")
+
+        bucket, name = self._get_bucket_name(url)
+
+        if 'r' in mode:
+            return await self._storage_client.get_object(bucket, name)
+
+        assert 'w' in mode
+        return await self._storage_client.insert_object(bucket, name)
+
+    async def mkdir(self, url: str) -> None:
+        pass
+
+    async def isfile(self, url: str) -> bool:
+        try:
+            bucket, name = self._get_bucket_name(url)
+            await self._storage_client.get_object_metadata(bucket, name)
+            return True
+        except aiohttp.ClientResponseError as e:
+            if e.status == 404:
+                return False
+            raise
+
+    async def isdir(self, url: str) -> bool:
+        bucket, name = self._get_bucket_name(url)
+        if not name.endswith('/'):
+            name = f'{name}/'
+        params = {
+            'delimiter': '/',
+            'includeTrailingDelimiter': 'true',
+            'maxResults': 1,
+            'prefix': name
+        }
+        data = await self._storage_client._list_objects(bucket, params=params)
+        assert 'prefixes' not in data or data['prefixes'] is None
```

Comment:
Is the prefixes in the case where you don't specify a delimiter?

Reply:
I think this code is wrong.  I need to test it on nested directories.

So I don't totally understand the Google API on this point, but list objects returns two things: the items you requested, and "prefixes".  See:

https://cloud.google.com/storage/docs/json_api/v1/objects/list

 > prefixes[] contains truncated object names for objects whose names contain delimiter after any prefix. Object names are truncated beyond the first applicable instance of the delimiter. If multiple objects have the same truncated name, duplicates are omitted.

It sounds like prefixes basically "subdirectories" of the requested prefix.  However, `maxResults` limits the combined items and prefixes:

 > Maximum combined number of entries in items[] and prefixes[] to return in a single page of responses. Because duplicate entries in prefixes[] are omitted, fewer total results may be returned than requested. The service uses this parameter or 1,000 items, whichever is smaller.

So I think I need to check prefixes and items here.

I will run some experiments and add some additional tests.

Reply:
OK, yeah, when you specify a delimiter, prefixes are essentially the subdirectories, and items are the files.  I modified the code to look for a prefix or an item, and added a subdirectory test.